homeassistant/components/subaru/config_flow.py

```diff
@@ -0,0 +1,179 @@
+            else:
+                if self.controller.is_pin_required():
+                    return await self.async_step_pin()
+                return self.async_create_entry(
+                    title=user_input[CONF_USERNAME], data=self.config_data
+                )
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_USERNAME,
+                        default=user_input.get(CONF_USERNAME) if user_input else "",
+                    ): str,
+                    vol.Required(
+                        CONF_PASSWORD,
+                        default=user_input.get(CONF_PASSWORD) if user_input else "",
+                    ): str,
+                    vol.Required(
+                        CONF_COUNTRY,
+                        default=user_input.get(CONF_COUNTRY)
+                        if user_input
+                        else COUNTRY_USA,
+                    ): vol.In([COUNTRY_CAN, COUNTRY_USA]),
+                }
+            ),
+            errors=error,
+        )
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get the options flow for this handler."""
+        return OptionsFlowHandler(config_entry)
+
+    async def validate_login_creds(self, data):
+        """Validate the user input allows us to connect.
+
+        data: contains values provided by the user.
+        """
+        websession = aiohttp_client.async_get_clientsession(self.hass)
+        now = datetime.now()
+        if not data.get(CONF_DEVICE_ID):
+            data[CONF_DEVICE_ID] = int(now.timestamp())
+        date = now.strftime("%Y-%m-%d")
+        device_name = "Home Assistant: Added " + date
+
+        self.controller = SubaruAPI(
+            websession,
+            username=data[CONF_USERNAME],
+            password=data[CONF_PASSWORD],
+            device_id=data[CONF_DEVICE_ID],
+            pin=None,
+            device_name=device_name,
+            country=data[CONF_COUNTRY],
+        )
+        _LOGGER.debug("Using subarulink %s", self.controller.version)
+        _LOGGER.debug(
+            "Setting up first time connection to Subuaru API.  This may take up to 20 seconds."
+        )
+        if await self.controller.connect():
+            _LOGGER.debug("Successfully authenticated and authorized with Subaru API")
+            self.config_data.update(data)
+
+    async def async_step_pin(self, user_input=None):
+        """Handle second part of config flow, if required."""
+        error = None
+        if user_input:
+            if self.controller.update_saved_pin(user_input[CONF_PIN]):
+                try:
+                    vol.Match(r"[0-9]{4}")(user_input[CONF_PIN])
+                    await self.controller.test_pin()
+                    _LOGGER.debug("PIN successfully tested")
+                    self.config_data.update(user_input)
```

Comment:
Move this out of the try block into the `else` block since its not ever going to throw

Reply:
updated